File: apps/mobile/lib/src/connector/state.dart
/// The connector's state, as the window receives it.
///
/// A mirror of `UiState.as_json` in
/// `apps/connector/src/tally_connector/ui/state.py`, which is why both files
/// carry the same field names in the same order.
///
/// Every field here is read defensively. The window and the connector ship in
/// one installer, but a machine that failed halfway through an update runs one
/// old half against one new half -- and a status window that throws on a
/// missing key is a status window that cannot report the very failure it is
/// looking at.
library;

/// Reads a string, whatever the connector actually sent.
String _text(Object? value) => value is String ? value : '';

int _count(Object? value) => value is int ? value : 0;

bool _flag(Object? value) => value is bool && value;

/// Reads a tri-state flag: true, false, or "not looked yet".
bool? _tristate(Object? value) => value is bool ? value : null;

Map<String, Object?> _object(Object? value) =>
    value is Map<String, Object?> ? value : const <String, Object?>{};

List<Map<String, Object?>> _rows(Object? value) {
  if (value is! List<Object?>) {
    return const <Map<String, Object?>>[];
  }
  return value.whereType<Map<String, Object?>>().toList(growable: false);
}

List<String> _strings(Object? value) {
  if (value is! List<Object?>) {
    return const <String>[];
  }
  return value.whereType<String>().toList(growable: false);
}

DateTime? _moment(Object? value) =>
    value is String ? DateTime.tryParse(value)?.toLocal() : null;

/// Everything the window draws.
class ConnectorState {
  const ConnectorState({
    required this.version,
    required this.uptime,
    required this.paired,
    required this.connectorId,
    required this.connectorName,
    required this.pairing,
    required this.backend,
    required this.tally,
    required this.account,
    required this.logDir,
  });

  factory ConnectorState.fromJson(Map<String, Object?> json) => ConnectorState(
        version: _text(json['version']),
        uptime: Duration(seconds: _count(json['uptime_seconds'])),
        paired: _flag(json['paired']),
        connectorId: _text(json['connector_id']),
        connectorName: _text(json['connector_name']),
        pairing: PairingState.fromJson(_object(json['pairing'])),
        backend: BackendState.fromJson(_object(json['backend'])),
        tally: TallyState.fromJson(_object(json['tally'])),
        account: AccountState.fromJson(_object(json['account'])),
        logDir: _text(json['log_dir']),
      );

  final String version;
  final Duration uptime;
  final bool paired;
  final String connectorId;
  final String connectorName;
  final PairingState pairing;
  final BackendState backend;
  final TallyState tally;
  final AccountState account;
  final String logDir;

  /// What the window leads with: a code to scan, or this machine's status.
  ///
  /// Keyed off the pairing code rather than off [paired], because the two
  /// disagree for a few seconds in both directions -- while a scan is being
  /// collected, and after a re-pair revokes this machine -- and the code on
  /// screen is the one of the two somebody is acting on.
  bool get isPairing => pairing.waiting;
}

/// The code on screen, while there is one.
class PairingState {
  const PairingState({
    required this.waiting,
    required this.payload,
    required this.secondsLeft,
    required this.detail,
  });

  factory PairingState.fromJson(Map<String, Object?> json) => PairingState(
        waiting: _flag(json['waiting']),
        payload: _text(json['payload']),
        secondsLeft: _count(json['seconds_left']),
        detail: _text(json['detail']),
      );

  final bool waiting;

  /// The QR's contents. Compared between polls to decide whether the code
  /// already drawn is still the right one, so the window never redraws a symbol
  /// that has not changed.
  final String payload;

  /// Recomputed by the connector on every read, never counted down here. A
  /// window ticking its own copy would keep promising time on a code the
  /// backend had already expired.
  final int secondsLeft;

  final String detail;
}

/// Whether TallyFlow is answering.
class BackendState {
  const BackendState({
    required this.url,
    required this.connected,
    required this.detail,
    required this.sessionId,
    required this.connectedFor,
  });

  factory BackendState.fromJson(Map<String, Object?> json) => BackendState(
        url: _text(json['url']),
        connected: _flag(json['connected']),
        detail: _text(json['detail']),
        sessionId: _text(json['session_id']),
        connectedFor: json['connected_seconds'] is int
            ? Duration(seconds: json['connected_seconds']! as int)
            : null,
      );

  final String url;
  final bool connected;
  final String detail;
  final String sessionId;

  /// How long this socket has been up. Null when it is not.
  final Duration? connectedFor;
}

/// Whether TallyPrime is answering, and what it has open.
class TallyState {
  const TallyState({
    required this.host,
    required this.port,
    required this.online,
    required this.companiesOpen,
  });

  factory TallyState.fromJson(Map<String, Object?> json) => TallyState(
        host: _text(json['host']),
        port: _count(json['port']),
        online: _tristate(json['online']),
        companiesOpen: _strings(json['companies_open']),
      );

  final String host;
  final int port;

  /// Null before the first probe. Not false: "we have not looked yet" and
  /// "TallyPrime is not answering" are different things to tell somebody
  /// standing in front of the machine, and the second one sends them to restart
  /// software that is running perfectly.
  final bool? online;

  final List<String> companiesOpen;
}

/// What the account looks like from here: names, roles and sync times.
///
/// Never a figure. A shop PC that could be asked for a balance over its own
/// loopback socket would be a read path into the books outside every check the
/// backend makes.
class AccountState {
  const AccountState({
    required this.organisation,
    required this.status,
    required this.companies,
    required this.users,
    required this.asOf,
  });

  factory AccountState.fromJson(Map<String, Object?> json) => AccountState(
        organisation: _text(json['organisation']),
        status: _text(json['status']),
        companies: _rows(json['companies'])
            .map(FedCompany.fromJson)
            .toList(growable: false),
        users: _rows(json['users'])
            .map(RosterUser.fromJson)
            .toList(growable: false),
        asOf: _moment(json['as_of']),
      );

  final String organisation;
  final String status;
  final List<FedCompany> companies;
  final List<RosterUser> users;

  /// When the roster last arrived. Null until one has.
  final DateTime? asOf;

  bool get isEmpty => companies.isEmpty && users.isEmpty;
}

/// A company this computer feeds.
class FedCompany {
  const FedCompany({
    required this.id,
    required this.name,
    required this.tallyName,
    required this.isActive,
    required this.lastSyncedAt,
  });

  factory FedCompany.fromJson(Map<String, Object?> json) => FedCompany(
        id: _text(json['id']),
        name: _text(json['name']),
        tallyName: _text(json['tally_name']),
        isActive: _flag(json['is_active']),
        lastSyncedAt: _moment(json['last_synced_at']),
      );

  final String id;
  final String name;

  /// What the company is called inside TallyPrime, when that differs from what
  /// the account calls it. The two drift apart the moment somebody renames one,
  /// and the difference is what explains an empty dashboard.
  final String tallyName;

  final bool isActive;

  /// Null means never synced -- which is "not synced yet", never "synced never"
  /// and never a zero.
  final DateTime? lastSyncedAt;
}

/// Somebody in the business who can see what this computer feeds.
class RosterUser {
  const RosterUser({
    required this.name,
    required this.email,
    required this.role,
    required this.hasAccess,
  });

  factory RosterUser.fromJson(Map<String, Object?> json) => RosterUser(
        name: _text(json['name']),
        email: _text(json['email']),
        role: _text(json['role']),
        hasAccess: _flag(json['has_access']),
      );

  final String name;
  final String email;
  final String role;

  /// Whether this person can see this computer's companies. A colleague on the
  /// account who cannot is a normal, deliberate state, so it is shown rather
  /// than filtered out.
  final bool hasAccess;
}

/// What came back from one of the four buttons.
class ActionResult {
  const ActionResult({required this.ok, required this.message});

  factory ActionResult.fromJson(Map<String, Object?> json) => ActionResult(
        ok: _flag(json['ok']),
        message: _text(json['message']),
      );

  final bool ok;
  final String message;
}

File: apps/connector/src/tally_connector/ui/state.py
"""What the local window shows, and how a click on it reaches the event loop.

Two threads meet here and neither may block the other. The connector's work
happens on an asyncio loop; the window polls a small threaded HTTP server from
its own process, because a shop PC should not have a web framework installed on
it to answer one status screen. So:

* **Reading** is a snapshot under a lock. The window never touches live objects,
  which is what stops a poll from being able to observe the session halfway
  through a reconnect.
* **Writing** is a coroutine posted onto the loop and waited for with a timeout.
  A button that hangs is worse than a button that says it could not do it -- the
  person pressing it is already looking at a machine they think is broken.

The bridge also satisfies the observer protocol in :mod:`tally_connector.session`,
which is how the session reports connection and Tally state without importing
anything about a user interface. That direction matters: the connector must run
perfectly with ``ui_enabled`` off, and it does, because nothing below this
module knows the window exists.
"""

from __future__ import annotations

import asyncio
import logging
import threading
import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any

from tally_core.protocol import Roster

logger = logging.getLogger(__name__)

#: Longest a button press may take before the window is told it failed. Chosen
#: against the slowest action -- a restart, which tears down a socket and stands
#: a new session up -- rather than against the fastest.
ACTION_TIMEOUT_SECONDS = 20.0

#: ``(payload) -> result``. Runs on the connector's event loop.
Action = Callable[[dict[str, Any]], Awaitable[dict[str, Any]]]


def _iso(value: datetime | None) -> str | None:
    return None if value is None else value.astimezone(UTC).isoformat()


@dataclass
class UiState:
    """Everything the window draws, in one object with one owner.

    Deliberately flat and JSON-shaped. The window is the only reader, it renders
    what it is given, and a snapshot that needed assembling from three live
    objects would be a snapshot that could disagree with itself.
    """

    version: str = ""
    started_at: float = field(default_factory=time.time)

    # -- pairing ---------------------------------------------------------
    paired: bool = False
    connector_id: str = ""
    connector_name: str = ""
    #: Set while this machine is waiting to be scanned. Its presence is what the
    #: window keys the whole pairing screen off.
    pairing_payload: str = ""
    #: When the code on screen stops working, as a wall-clock moment rather than
    #: a countdown. A stored "seconds left" would be whatever it was when the
    #: claim was opened, and the window would cheerfully promise fourteen minutes
    #: for the whole fourteen minutes.
    pairing_expires_at: float | None = None
    pairing_detail: str = ""

    # -- backend ---------------------------------------------------------
    backend_url: str = ""
    backend_connected: bool = False
    backend_detail: str = ""
    session_id: str = ""
    connected_since: float | None = None

    # -- Tally -----------------------------------------------------------
    tally_host: str = ""
    tally_port: int = 0
    #: ``None`` before the first probe. Not ``False``: "we have not looked yet"
    #: and "TallyPrime is not answering" are different things to tell somebody
    #: standing in front of the machine, and the second one sends them to
    #: restart software that is running perfectly.
    tally_online: bool | None = None
    companies_open: list[str] = field(default_factory=list)

    # -- roster ----------------------------------------------------------
    organisation: str = ""
    org_status: str = ""
    companies: list[dict[str, Any]] = field(default_factory=list)
    users: list[dict[str, Any]] = field(default_factory=list)
    roster_at: datetime | None = None

    log_dir: str = ""

    def as_json(self) -> dict[str, Any]:
        return {
            "version": self.version,
            "uptime_seconds": int(time.time() - self.started_at),
            "paired": self.paired,
            "connector_id": self.connector_id,
            "connector_name": self.connector_name,
            "pairing": {
                "waiting": bool(self.pairing_payload),
                "payload": self.pairing_payload,
                "seconds_left": (
                    0
                    if self.pairing_expires_at is None
                    else max(0, int(self.pairing_expires_at - time.time()))
                ),
                "detail": self.pairing_detail,
            },
            "backend": {
                "url": self.backend_url,
                "connected": self.backend_connected,
                "detail": self.backend_detail,
                "session_id": self.session_id,
                "connected_seconds": (
                    None
                    if self.connected_since is None
                    else int(time.time() - self.connected_since)
                ),
            },
            "tally": {
                "host": self.tally_host,
                "port": self.tally_port,
                "online": self.tally_online,
                "companies_open": list(self.companies_open),
            },
            "account": {
                "organisation": self.organisation,
                "status": self.org_status,
                "companies": list(self.companies),
                "users": list(self.users),
                "as_of": _iso(self.roster_at),
            },
            "log_dir": self.log_dir,
        }


class UiBridge:
    """The window's view of the connector, and its handful of levers."""

    def __init__(self, *, version: str, log_dir: str = "") -> None:
        self._lock = threading.Lock()
        self._state = UiState(version=version, log_dir=log_dir)
        self._loop: asyncio.AbstractEventLoop | None = None
        self._actions: dict[str, Action] = {}

    # -- wiring ----------------------------------------------------------

    def bind(self, loop: asyncio.AbstractEventLoop) -> None:
        """Attach the loop that actions will be posted onto.

        Called from the loop's own thread when the connector starts. Before it,
        the window still renders -- it just reports that nothing can be pressed
        yet, which is true.
        """
        self._loop = loop

    def register(self, name: str, action: Action) -> None:
        self._actions[name] = action

    # -- reading ---------------------------------------------------------

    def snapshot(self) -> dict[str, Any]:
        with self._lock:
            return self._state.as_json()

    def pairing_payload(self) -> str:
        with self._lock:
            return self._state.pairing_payload

    def update(self, **fields: Any) -> None:
        """Set fields on the state. Unknown names are ignored on purpose.

        This is called from the connector's hot paths -- a status transition, a
        roster arriving -- and an attribute typo there must cost a wrong window,
        never a dropped session.
        """
        with self._lock:
            for key, value in fields.items():
                if hasattr(self._state, key):
                    setattr(self._state, key, value)
                else:
                    logger.debug("ignoring unknown UI field %r", key)

    # -- the session observer protocol -----------------------------------

    def connection_changed(
        self, *, connected: bool, session_id: str = "", detail: str = ""
    ) -> None:
        self.update(
            backend_connected=connected,
            session_id=session_id,
            backend_detail=detail,
            connected_since=time.time() if connected else None,
        )

    def tally_changed(
        self, *, online: bool | None, companies_open: list[str] | None = None
    ) -> None:
        fields: dict[str, Any] = {"tally_online": online}
        # An empty list from a heartbeat means "this frame carried no company
        # list", not "no company is open" -- overwriting on it would blank the
        # one line that explains a stale dashboard.
        if companies_open:
            fields["companies_open"] = list(companies_open)
        self.update(**fields)

    def roster_received(self, roster: Roster) -> None:
        fields: dict[str, Any] = {
            "organisation": roster.organisation,
            "org_status": roster.org_status,
            "companies": [company.model_dump(mode="json") for company in roster.companies],
            "users": [user.model_dump(mode="json") for user in roster.users],
            "roster_at": datetime.now(UTC),
        }
        # Only when the roster actually carries one. Reading the current value
        # to fall back on would mean taking the lock inside the argument list of
        # a call that takes it again -- which works today only because the two
        # do not overlap, and is a deadlock the first time somebody tidies it.
        if roster.connector_name:
            fields["connector_name"] = roster.connector_name
        self.update(**fields)

    # -- writing ---------------------------------------------------------

    def invoke(self, name: str, payload: dict[str, Any] | None = None) -> dict[str, Any]:
        """Run a registered action on the event loop and wait for its answer.

        Called from an HTTP worker thread. Every failure comes back as a result
        the window can render, never as an exception across the thread boundary --
        a traceback in a status window is not an answer to "why is my dashboard
        empty?".
        """
        action = self._actions.get(name)
        loop = self._loop
        if action is None:
            return {"ok": False, "message": f"{name} is not something this build can do."}
        if loop is None or loop.is_closed():
            return {"ok": False, "message": "The connector is still starting up."}

        future = asyncio.run_coroutine_threadsafe(action(payload or {}), loop)
        try:
            return future.result(timeout=ACTION_TIMEOUT_SECONDS)
        except TimeoutError:
            # Deliberately not cancelled. A restart that is merely slow will
            # still finish, and cancelling it halfway would leave the connector
            # in the one state nothing else recovers from.
            return {
                "ok": False,
                "message": "That is taking longer than expected. Check back in a moment.",
            }
        except Exception as exc:  # noqa: BLE001 - the window must always get an answer
            logger.warning("UI action %s failed", name, exc_info=True)
            return {"ok": False, "message": str(exc)}
